storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,359 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"regexp"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/wal"
+	fsnotify "gopkg.in/fsnotify/fsnotify.v1"
+)
+
+var (
+	segmentRegex              = regexp.MustCompile(`\d{8}`)
+	checkpointRegex           = regexp.MustCompile(`checkpoint.\d{6}`)
+	watcherSamplesRecordsRead = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "samples_records_read",
+			Help:      "Number of samples records read by the WAL watcher from the WAL.",
+		},
+		[]string{queue},
+	)
+	watcherSeriesRecordsRead = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "series_records_read",
+			Help:      "Number of series records read by the WAL watcher from the WAL.",
+		},
+		[]string{queue},
+	)
+	currentSegmentMetric = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "current_segment",
+			Help:      "Current segment the WAL watcher is reading records from.",
+		},
+		[]string{queue},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(watcherSamplesRecordsRead)
+	prometheus.MustRegister(watcherSeriesRecordsRead)
+	prometheus.MustRegister(currentSegmentMetric)
+}
+
+type WriteTo interface {
+	Append([]tsdb.RefSample) bool
+	StoreSeries([]tsdb.RefSeries)
+	ClearSeries()
+	Name() string
+}
+
+// WALWatcher watches the TSDB WAL for a given WriteTo.
+type WALWatcher struct {
+	writer         WriteTo
+	logger         log.Logger
+	walDir         string
+	watcher        *fsnotify.Watcher
+	currentSegment int
+	lastCheckpoint string
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	quit   chan struct{}
+}
+
+// NewWALWatcher creates a new WAL watcher for a given WriteTo.
+func NewWALWatcher(logger log.Logger, writer WriteTo, walDir string) *WALWatcher {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	w := &WALWatcher{
+		logger: logger,
+		writer: writer,
+		walDir: path.Join(walDir, "wal"),
+		ctx:    ctx,
+		cancel: cancel,
+		quit:   make(chan struct{}),
+	}
+	return w
+}
+
+func (w *WALWatcher) Start() {
+	go w.runWatcher()
+}
+
+func (w *WALWatcher) Stop() {
+	close(w.quit)
+}
+
+func (w *WALWatcher) readSeriesRecords(r *wal.LiveReader) {
+	var (
+		dec    tsdb.RecordDecoder
+		series []tsdb.RefSeries
+	)
+
+	for {
+		err := r.Next()
+		if err != nil {
+			if err == io.EOF {
```

Comment:
Collapse this into one line `if err == io.EOF`.

You continue if err == that isn't EOF.

Reply:
Discussed: if this only returns EOF, probably want type to be bool?